File: Codes/2300-successful-pairs-of-spells-and-potions.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-11-10 14:17:25
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-11-10 14:23:41
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

typedef long long ll;
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, ll success) {
        sort(potions.begin(), potions.end());
        for (int& t : spells) {
            ll toFind = success / t;
            if (toFind * t < success) {
                toFind++;
            }
            t = potions.end() - lower_bound(potions.begin(), potions.end(), toFind);
        }
        return spells;
    }
};

File: Codes/2300-successful-pairs-of-spells-and-potions.py
'''
Author: LetMeFly
Date: 2023-11-10 14:25:21
LastEditors: LetMeFly
LastEditTime: 2023-11-10 14:28:30
'''
from typing import List
from bisect import bisect_left

class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        for i in range(len(spells)):
            toFind = success // spells[i]
            if toFind * spells[i] < success:
                toFind += 1
            spells[i] = len(potions) - bisect_left(potions, toFind)
        return spells
